torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp

```diff
@@ -116,6 +118,59 @@ static bool needsGradientInProfilingMode(Block* b) {
   return false;
 }
 
+bool guardDifferentiableGraph(Node* dnode) {
+  auto gi = dnode->g(attr::Subgraph)->inputs();
+  bool all_inputs_seen = true;
+  for (size_t i = 0; i < gi.size(); i++) {
+    auto ty = gi[i]->type()->cast<TensorType>();
+    if (ty) {
+      auto n = gi[i]->uses().at(0).user;
+      auto dni = dnode->inputs().at(i);
+      GRAPH_DEBUG("found first user of ", i, " as ", *n);
+      if (n->kind() == prim::profile) {
+        GRAPH_DEBUG(
+            "setting input ", i, " to type ", *n->ty(attr::profiled_type));
+        dni->setType(n->ty(attr::profiled_type));
+      } else if (dni->node()->kind() == prim::DifferentiableGraph) {
```

Comment:
OMG

Reply:
Yeah. Eventually, I think the best fix would be to 1) re-profile the graph inside the differentiable graph (or do we do that already when using the differentiable graph operator?) or 2) apply the type annotations before creating autodiff graphs. So I understood @eellison that this isn't good to do at the moment, but I haven't understood well enough what is unsafe about it to see if that could be fixed.

Reply:
1) is an easy enough - see https://github.com/pytorch/pytorch/pull/47434. 

2) yea is a little weird bc it gets away from our "fusers decide what to guard on" logic, and im also a little worried about proving what properties in the differentiable graph are guaranteed. For instance, if an Op takes in a NumberType, the output type could change if in one iteration the input is an Integer and the next a Float. I'm not saying that's particularly common, but it's just one example where strictly guarding on Tensor inputs does not guarantee consistent tensor types.